artifacts/reader.py

```diff
@@ -224,6 +208,57 @@ def _ReadSupportedOS(self, definition, definition_object, name):
 
     definition_object.supported_os = supported_os
 
+  def _ReadSources(self, artifact_definition, artifact_object):
+    """Reads the artifact definition sources
+
+    Args:
+      artifact_definition: the artifact definition as a dict.
+      artifact_object: the artifact definition object (instance of
+                           ArtifactDefinition).
+
+    Raises:
+      FormatError: if the type indicator is not set or unsupported,
+                   or if required attributes are missing.
+    """
+
+    name = artifact_definition.get(u'name', None)
+    sources = artifact_definition.get(u'sources')
+
+    if not sources:
+      raise errors.FormatError(
+          u'Invalid artifact definition: {0} missing sources.'.format(name))
+
+    for source_definition in sources:
+      type_indicator = source_definition.get(u'type', None)
+      if not type_indicator:
+        raise errors.FormatError(
+            u'Invalid artifact definition: {0} source type.'.format(name))
+
+      if type_indicator not in self.type_indicators:
+        raise errors.FormatError(u'Unsupported type indicator: {0}.'.format(
+            type_indicator))
+
+      attributes = source_definition.get(u'attributes', None)
+
+      try:
+        source_object = artifact_object.AppendSource(type_indicator, attributes)
```

Comment:
I opt source_type (as the original) instead of source_object or source_type_object in case of.

>  I was mostly trying to keep it all separate due to making more direct use of the dicts in my use cases, but I will rename to whatever convention is desired.

Agree that the idea is good, but you overloaded artifact_definition by doing so, if you had picked a different name e.g. artifact_definition_dict or artifact_definition_values you would have reached your goal without overloading terminology. And then in the process need to overload more terminology ;)